lightning-background-processor/src/lib.rs

```diff
@@ -866,7 +866,7 @@ mod tests {
 		fn disconnect_socket(&mut self) {}
 	}
 
-	type ChannelManager = channelmanager::ChannelManager<Arc<ChainMonitor>, Arc<test_utils::TestBroadcaster>, Arc<KeysManager>, Arc<KeysManager>, Arc<KeysManager>, Arc<test_utils::TestFeeEstimator>, Arc<DefaultRouter< Arc<NetworkGraph<Arc<test_utils::TestLogger>>>, Arc<test_utils::TestLogger>, Arc<Mutex<TestScorer>>>>, Arc<test_utils::TestLogger>>;
+	type ChannelManager = channelmanager::ChannelManager<Arc<ChainMonitor>, Arc<test_utils::TestBroadcaster>, Arc<KeysManager>, Arc<KeysManager>, Arc<KeysManager>, Arc<test_utils::TestFeeEstimator>, Arc<DefaultRouter< Arc<NetworkGraph<Arc<test_utils::TestLogger>>>, Arc<test_utils::TestLogger>, Arc<Mutex<TestScorer>>, i32, TestScorer>>, Arc<test_utils::TestLogger>>;
```

Comment:
Wierdly, this part still has an issue with getting past compiler checks. im sure im missing something obvious here. im not quite sure why it found a `MutexGuard<'_, tests::TestScorer, >` when i did pass in a `Mutex` wrapped in `Arc`

```
error[E0271]: type mismatch resolving `for<'a> <Mutex<TestScorer> as LockableScore<'a>>::Locked == TestScorer`
   --> lightning-background-processor/src/lib.rs:877:9
    |
877 |         node: Arc<ChannelManager>,
    |               ^^^^^^^^^^^^^^^^^^^ expected struct `TestScorer`, found struct `MutexGuard`
    |
    = note: expected struct `tests::TestScorer`
               found struct `MutexGuard<'_, tests::TestScorer, >`
note: required by a bound in `DefaultRouter`
   --> /Users/henghonglee/lightning/hh/rust-lightning/lightning/src/routing/router.rs:38:40
    |
38  |     S::Target: for <'a> LockableScore<'a, Locked = Sc>,
    |                                           ^^^^^^^^^^^ required by this bound in `DefaultRouter`
```

Reply:
i still dont really understand how HRTB works here. 

I get now why the compiler is looking for `MutexGuard<'_, tests::TestScorer, >`, because its taking Mutex<T = TestScorer> and getting its locked value as in here

```
impl<'a, T: 'a + Score> LockableScore<'a> for Mutex<T> {
	type Locked = MutexGuard<'a, T>;

	fn lock(&'a self) -> MutexGuard<'a, T> {
		Mutex::lock(self).unwrap()
	}
}
```

however i because of the HRTB `for <'a>` in LockableScore im once again finding myself trying to define `<'a>` from a place where it isnt initialized. its unclear to me why the compiler doesnt infer the lifetimes to be a generic

```
type ChannelManager = channelmanager::ChannelManager<
	Arc<ChainMonitor>,
	Arc<test_utils::TestBroadcaster>,
	Arc<KeysManager>,
	Arc<KeysManager>,
	Arc<KeysManager>,
	Arc<test_utils::TestFeeEstimator>,
	Arc<DefaultRouter<
			Arc<NetworkGraph<Arc<test_utils::TestLogger>>>,
			Arc<test_utils::TestLogger>,
			Arc<Mutex<TestScorer>>,
			i32,
			MutexGuard<'static, TestScorer>,
	>>,
	Arc<test_utils::TestLogger>>;
```

furthur trying using `Sc = MutexGuard<'static, TestScorer>` as above i get the error below
```
mismatched types
expected trait `LockableScore<'a>`
   found trait `LockableScore<'_>`
```

Passing in the lifetimes into `type ChannelManager<'a>` also doesnt work.

Reply:
solved